### ISSUE TITLE
Possible error in the "eval" function in "08-rcfae.scala"

### ISSUE THREAD

**Issue by rzhxeo** [klauso/PLT2013#3]
I think there is an error in the "eval" function.
In `case App...` the environment is copied:
`closureEnv + (f.param -> eval(a,env))`
It's no longer the same reference as `mutableenv`.

Example:

``` scala
var func = Fun('n, If0('n, 0, App('func, Add('n, -1))))
var works = Letrec('func, func, App('func, 1))
var fails = Letrec('func, App(Fun('notUsed, func), 0), App('func, 1))
assert(eval(works, Map()) == NumV(0)) // ok
assert(eval(fails, Map()) == NumV(0)) // error
```

**Comment by reflectionalist** [comment]
Thanks for opening the issue.  I appreciate your exploration.

> In case App... the environment is copied:
> `closureEnv + (f.param -> eval(a,env))`
> It's no longer the same reference as `mutableenv`.

~~Now it is clearer what you meant by "the environment is copied".  You mean when `eval` evaluates `App(Fun('notUsed, func), 0)`, the environment at that point referenced by `mutableenv` is passed to a recursive call of `eval` _by value_, not _by reference_.  Indeed, because Scala is by default _pass-by-value_.  Although _pass-by-reference_ can be simulated, see [here](http://stackoverflow.com/questions/6896118/basic-scala-oop-question-pass-by-reference).~~  This is a mis-comment.  Objects in OOLs are by default passed by reference.

> I think there is an error in the "eval" function.

I do not consider this an error.  Though It is indeed a limitation.  Although not stated explicitly in the lecture notes, the expression whose value will be bound to the variable name introduced by `Letrec` is supposed to be one that _directly_ creates an anonymous function in which the variable name can be referenced.  In other words, the expression should start with `Fun`.  Your example of `fails` goes beyond this assumption.  But I agree that in practice this limitation would be annoying.  If you really want to support expression to its full extent in the second position of `Letrec`, you can modify the evaluator following the trick linked above to do so.  I just tried, it works.  But for pedagogical purpose, I think it may be better to admit this limitation.

@klauso and @yfcai: What is your opinion about this?  If we agree to admit this limitation, I can easily fix the hole that allows a definition like `fails` by shrinking the type assigned to the second argument of `Letrec`.  Currently, it is `Exp`.  We only need to change it to `Fun`.

**Comment by klauso** [comment]
Thanks for the very good and insightful comment!

I think you are right: The "fails" example should work. I erroneously expected "+" to delegate to closureEnv instead of copying closureEnv. We need a variant of "+" here that does not copy closureEnv.

One way to achieve the desired purpose is this variant of the "App" case:

```
case ClosureV(f,closureEnv) => {
  val av = eval(a,env)
  val newEnv = new DefaultMap[Symbol,Value] {
    def get(key: Symbol) = if (key == f.param) Some(av) else closureEnv.get(key)
    def iterator = Iterator.single((f.param,av)) ++ closureEnv.iterator
  }
  eval(f.body, newEnv)
}
```

Presumably there is a more elegant way to achieve the same thing.

**Comment by yfcai** [comment]
This is an absolutely amazing find! It demonstrates the incredible difficulty of using circularity correctly, as well as how an innocuous language feature can break some other unrelated features. In this case, introducing `Letrec` broke `App`.

Of course, everybody is right: @rzhxeo reasoned correctly that it is vital to store `mutableenv` in the value bound to the recursive name in the environment. Yi's suggestion of having `Fun` in the second argument of `Letrec` would achieve exactly that: The `Fun` would evaluate to a closure with `mutableenv` in it.

Attached are 2 alternatives to fix `Letrec` and leave `App` untouched.

**Comment by klauso** [comment]
Eta expansion is not sound when the expression that is eta-expanded is not a function. For instance, I assume something like letrec x = 1 in x+1 wouldn't work. 

The thing I don't like about the lazy solution is that it would change the language semantics. For instance, letrec omega = (\x. (x x) \x. (x x)) should diverge in a call-by-value language, but from a cursory look it wouldn't in your variant.

I think the two best options to fix the problem are to either restrict the language (as Yi proposed) or to not copy the mutable environment (as in my proposal). Restricting the language would be a little sad, though, since it would rule out useful programs. 

I also looked into Krishnamurthi's book, see p.101/102 here http://cs.brown.edu/~sk/Publications/Books/ProgLangs/2007-04-26/plai-2007-04-26.pdf

What he does in the App case corresponds somewhat to my proposal. He uses boxes in the environment and then changes the box. We could do that as well, but it's also rather complicated.

I think the best solution would be to introduce a generic variant of Map.+ that maintains the reference to the extended map instead of copying it into a new map.

**Comment by yfcai** [comment]
> Eta expansion is not sound when the expression that is eta-expanded is not a function. For instance, I assume something like letrec x = 1 in x+1 wouldn't work.

Ah, that's right, thanks for pointing it out.

> The thing I don't like about the lazy solution is that it would change the language semantics. For instance, letrec omega = (\x. (x x) \x. (x x)) should diverge in a call-by-value language, but from a cursory look it wouldn't in your variant.

Sorry, I overlooked this aspect. It's possible to fix (715dd78bad233d8d4003db7538d9de7b3c1211e1). Not sure if a solution by restricted laziness is better, but at least we've some alternative to changing the environment representation.
